core/miner/storage_fsm/impl/precommit_batcher_impl.cpp

```diff
@@ -0,0 +1,137 @@
+
+    OUTCOME_TRY(head, api_->ChainHead());
+    OUTCOME_TRY(minfo, api_->StateMinerInfo(miner_address_, head->key));
+
+    PreCommitBatch::Params params = {};
+    for (auto &data : batchStorage) {
+      mutualDeposit += data.second.deposit;
+      params.sectors.push_back(data.second.pci);
+    }
+    // TODO: goodFunds = mutualDeposit + MaxFee; -  for checking payable
+    if (params.sectors.size() != 0) {
+      OUTCOME_TRY(encodedParams, codec::cbor::encode(params));
+
+      auto maybe_signed = api_->MpoolPushMessage(
+          vm::message::UnsignedMessage(
+              miner_address_,
+              minfo.worker,  // TODO: handle worker
+              0,
+              mutualDeposit,
+              {},
+              {},
+              vm::actor::builtin::v0::miner::PreCommitSector::Number,
+              MethodParams{encodedParams}),
+          kPushNoSpec);
+
+      if (maybe_signed.has_error()) {
+        return maybe_signed.error(); //TODO: maybe logs
+      }
+
+      mutualDeposit = 0;
+      batchStorage.clear();
+    }
+    cutoffStart = std::chrono::system_clock::now();
+    return outcome::success();
+  }
+
+  outcome::result<void> PreCommitBatcherImpl::forceSend() {
+    auto error = sendBatch();
+    assert(not error.has_error());
+    handle_.reschedule(maxDelay);
+    return outcome::success();
+  }
+
+  void PreCommitBatcherImpl::getPreCommitCutoff(ChainEpoch curEpoch,
+                                                const SectorInfo &si) {
+    primitives::ChainEpoch cutoffEpoch =
+        si.ticket_epoch + static_cast<int64_t>(fc::kEpochsInDay + 600);
+    ChainEpoch startEpoch{};
+    for (auto p : si.pieces) {
+      if (!p.deal_info) {
+        continue;
+      }
+      startEpoch = p.deal_info->deal_schedule.start_epoch;
+      if (startEpoch < cutoffEpoch) {
+        cutoffEpoch = startEpoch;
+      }
+    }
+    if (cutoffEpoch <= curEpoch) {
+      auto maybe_error = forceSend();
+    } else {
+      Ticks tempCutoff = toTicks(std::chrono::seconds((cutoffEpoch - curEpoch)
+                                                      * kEpochDurationSeconds));
+      if ((closestCutoff
+               - toTicks(std::chrono::duration_cast<std::chrono::seconds>(
+                   std::chrono::system_clock::now() - cutoffStart))
+           > tempCutoff)) {
+        cutoffStart = std::chrono::system_clock::now();
+        handle_.reschedule(tempCutoff);
+        closestCutoff = tempCutoff;
+      }
+    }
+  }
+
+  outcome::result<void> PreCommitBatcherImpl::addPreCommit(
+      SectorInfo secInf,
+      primitives::TokenAmount deposit,
+      api::SectorPreCommitInfo pcInfo) {
+    std::unique_lock<std::mutex> locker(mutex_, std::defer_lock);
+    OUTCOME_TRY(head, api_->ChainHead());
+
+    getPreCommitCutoff(head->epoch(), secInf);
+    auto sn = secInf.sector_number;
+    batchStorage[sn] = preCommitEntry(std::move(deposit), std::move(pcInfo));
+    return outcome::success();
+  }
+}  // namespace fc::mining
```

Comment:
Please, configure your IDE to add empty lines on file end.

Reply:
fixed